Proofs/Experiments/MemoryAliasing.lean

```diff
@@ -33,14 +121,171 @@ theorem mem_automation_test_2
+    (hlegal : mem_legal' src_addr 16) :
+    Memory.read_bytes 16 src_addr (Memory.write_bytes 16 src_addr val mem) =
+     val.extractLsBytes 0 16  := by
+  simp_mem
+  · -- ⊢ val.extractLsBytes (src_addr.toNat - src_addr.toNat) 16 = val.extractLsBytes 0 16
+    congr
+    simp
+
+theorem test_2 {val : BitVec _}
+    (hlegal : mem_legal' src_addr 16) :
+    Memory.read_bytes 6 (src_addr + 10) (Memory.write_bytes 16 src_addr val mem) =
+    val.extractLsBytes 10 6 := by
+  simp_mem
+  have : ((src_addr + 10).toNat - src_addr.toNat) = 10 := by bv_omega'
+  rw [this]
+
+/--
+TODO(@bollu): the definition of overlap doesn't seem to do well with zero width!
+That's pretty interesting. I wonder if we ever need this though.
+-/
+theorem test_write_zero (hlegalw : mem_legal' write_addr 0)
+    (hlegalr : mem_legal' read_addr read_n) :
+    Memory.read_bytes read_n read_addr (Memory.write_bytes 0 write_addr write_val mem) =
+    mem.read_bytes read_n read_addr := by
+  simp_mem
+  sorry
+
+end ReadOverlappingWrite
+
+/- We check that we correctly visit the expression tree, both for binders,
+and for general walking. -/
+namespace ExprVisitor
+
+/-- Check that we correctly go under binders -/
+theorem test_quantified_1 {val : BitVec (16 * 8)}
+    (hlegal : mem_legal' 0 16) : ∀ (_irrelevant : Nat),
+    Memory.read_bytes 16 0 (Memory.write_bytes 16 0 val mem) =
+     val.extractLsBytes 0 16  := by
+  simp_mem
+  simp
+
+/--
+info: 'ExprVisitor.test_quantified_1' depends on axioms: [propext, Classical.choice, Quot.sound]
+-/
+#guard_msgs in #print axioms test_quantified_1
+
+/-- Check that we correctly walk under applications. -/
+theorem test_app_1 {val : BitVec (16 * 8)}
+    (hlegal : mem_legal' 0 16) (f : BitVec _ → Nat) :
+    f (Memory.read_bytes 16 0 (Memory.write_bytes 16 0 val mem)) =
+    f (val.extractLsBytes 0 16)  := by
+  simp_mem
+  simp
+/-- info: 'ExprVisitor.test_app_1' depends on axioms: [propext, Classical.choice, Quot.sound] -/
+#guard_msgs in #print axioms test_app_1
+
+/--
+Check that we correctly walk under applications (`f <walk inside>`)
+and binders (`∀ f, <walk inside>`) simultaneously.
+-/
+theorem test_quantified_app_1 {val : BitVec (16 * 8)}
+    (hlegal : mem_legal' 0 16) : ∀ (f : BitVec _ → Nat),
+    f (Memory.read_bytes 16 0 (Memory.write_bytes 16 0 val mem)) =
+    f (val.extractLsBytes 0 16)  := by
+  simp_mem
+  simp
+
+/--
+info: 'ExprVisitor.test_quantified_app_1' depends on axioms: [propext, Classical.choice, Quot.sound]
+-/
+#guard_msgs in #print axioms test_quantified_app_1
+
+/--
+Check that we correctly simplify the *types* of binders as well.
+Here, the bound variable `P` has *type* that involves a memory read.
+We make sure that we simplify these as well.
+-/
+theorem test_quantified_app_2 {val : BitVec (16 * 8)}
+    (_hlegal : ∀ (addr : Nat), mem_legal' addr 16)
+    (f : _ → Nat) :
+    f (∀ (_P : Memory.read_bytes 16 0 (Memory.write_bytes 16 0 val mem) = irrelevant), Nat) =
```

Comment:
Add a test case with an arrow of the same example to check that it's also a `forallE`.